omniledger/Contracts.md

```diff
@@ -0,0 +1,191 @@
+Navigation: [DEDIS](https://github.com/dedis/doc/tree/master/README.md) ::
+[Cothority](https://github.com/dedis/cothority/tree/master/README.md) ::
+[Building Blocks](https://github.com/dedis/cothority/tree/master/BuildingBlocks.md) ::
+[OmniLedger](README.md) ::
+Contracts and Instances
+
+# Contracts and Instances
+
+A contract in omniledger is similar to a smart contract in Ethereum, except that
+it is pre-compiled in the code and all nodes need to have the same version of
+the contract available in order to reach consensus.
+
+A contract can spawn new instances that are tied to another contract type. All
+instances are stored in the global state of omniledger. Every instance points
+to exactly one contract. An easy interpretation is to think of a contract as
+a class and the instance as an object instantiated from that class. Or, in go,
+as the instance being the values of the `struct` and the contract being all
+methods defined on that `struct`.
+
+## Authorizations
+
+Authorizations are handled using Darcs. Each Darc has a set of rules that define
+a pair of action / expression that need to be fulfilled to execute any instruction
+on the instances governed by that Darc.
+
+A Darc is always stored with the following `InstanceID`:
+
+`[32]byte{DarcBaseID}[32]byte{0}`
+
+All instances spawned by that Darc (except for other Darcs) will share the same
+first half of the `InstanceID`. The second half is randomly chosen by the contract
+having spawned this new instance.
+
+Given the following instruction sent by a client (some fields are omitted for
+clarity):
+
+- `InstanceID`: `[32]byte{GenesisDarc}[32]byte{0x01}`
+- `Invoke`:
+  - `Command`: `Update`
+  - `Args`: `{"Roster": NewRoster}`
+- `Signatures`: `[Sig1]`
+
+omniledger will do the following:
+
+1. find the Darc instance by looking at the first 32 bytes of the `InstanceID` given in
+the instruction, here the `GenesisDarc`, and adding 32 x 0x00 bytes
+2. create a `DarcRequest` using the `InstanceID` and the `Args`
+3. verify the request corresponds to the expression of the `Invoke_Update` rule
+in the Darc instance found in 1.
+
+## Contract Arguments
+
+A contract is immutable and has the following signature in go:
```

Comment:
Well, immutable until I type "go build" again. I don't think immutable is quite the word we want here...

s/go/Go/